File: cesta/SergioDeIscar-cesta-java/SergioDeIscar-cesta-java/src/main/java/org/example/models/Producto.java
package org.example.models;

public final class Producto {
    private String nombre;
    private Float precio;
    private int stock;

    public Producto(String nombre, Float precio, int stock) {
        if (nombre == null || nombre.isEmpty())
            throw new IllegalArgumentException("El nombre no puede ser nulo o vacío");
        if (precio <= 0)
            throw new IllegalArgumentException("El precio debe ser mayor a 0");
        if (stock <= 0)
            throw new IllegalArgumentException("El stock debe ser mayor a 0");
        setNombre(nombre);
        setPrecio(precio);
        setStock(stock);
    }

    //region Getters
    public String getNombre() {
        return nombre;
    }
    public Float getPrecio() {
        return precio;
    }
    public int getStock() {
        return stock;
    }
    //endregion
    //region Setters
    public void setNombre(String nombre) {
        if (nombre == null || nombre.isEmpty())
            throw new IllegalArgumentException("El nombre no puede ser nulo o vacío");
        this.nombre = nombre;
    }
    public void setPrecio(Float precio) {
        if (precio <= 0)
            throw new IllegalArgumentException("El precio debe ser mayor a 0");
        this.precio = precio;
    }
    public void setStock(int stock) {
        if (stock <= 0)
            throw new IllegalArgumentException("El stock debe ser mayor a 0");
        this.stock = stock;
    }
    //endregion

    @Override
    public String toString() {
        return "Producto -> Nombre: " + this.nombre + ", Precio -> " + this.precio + ", Stock -> " + this.stock;
    }
}

File: cesta/SergioDeIscar-cesta-kotlin/SergioDeIscar-cesta-kotlin/src/main/kotlin/models/Producto.kt
package factura.models

data class Producto(
    val nombre: String, private var precio: Float){
    var stock: Int = 0
        private set

    constructor(name: String, precio: Float, stock: Int): this(name, precio){
        check(name.isNotEmpty() && name.isNotBlank())
        check(precio > 0)
        SetStock(stock)
    }

    fun SetStock(value: Int){
        require(value >= 0)
        stock = value
    }

    fun SetPrecio(value: Float){
        require(value > 0)
        precio = value
    }

    fun GetPrecio(): Float{
        return precio
    }

    override fun toString(): String {
        return "Producto -> Nombre: ${this.nombre}, Precio -> ${this.precio}, Stock -> ${this.stock}"
    }
}